### PR TITLE
Add message queue trigger support

### PR DESCRIPTION
**Pull request by life1347**
This PR aims to add message queue support for #194 and it's not ready to merge yet. 

For people who want to try this feature
1. build new fission-bundle image & fission cli
2. apply `fission.yaml` to create new deployment (mqtrigger, nats-streaming) and service (nats-streaming)
3. follow fission cli instruction to create a message queue trigger 

Here is the demonstration video: https://youtu.be/mzjf_5fOrD8

Currently, message queue trigger only supports nats-streaming. CloudPubSub is in planning. The feature of sending function response back to event publisher will be implemented when the design of #217 is confirmed and merged.

Since it's still in development, any suggestion is welcome :)

### CODE REVIEW

File: mqtrigger/messageQueue/nats.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+			} else {
+				log.Infof("MessageQueue trigger %s deleted", trigger.Name)
+			}
+		}
+	}
+}
+
+func isTopicValidForNats(topic string) bool {
+	// nats-streaming does not support wildcard channl.
+	return nsUtil.IsSubjectValid(topic, false)
```
Comment (yqf3139) [inline mqtrigger/messageQueue/nats.go]: Compile failure. I think the `nsUtil.IsSubjectValid` just changed its arguments.

File: mqtrigger/messageQueue/nats.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+func (nats *Nats) add(trigger fission.MessageQueueTrigger) error {
+	subj := trigger.Topic
+
+	if !isTopicValidForNats(subj) {
+		return errors.New(fmt.Sprintf("Not a valid topic: %s", trigger.Topic))
+	}
+
+	handler := func(msg *ns.Msg) {
+		headers := map[string]string{
+			"X-Fission-Timer-Name": trigger.Function.Name,
```
Comment (life1347) [inline mqtrigger/messageQueue/nats.go]: yes, I noticed this before, it will be fixed in the later commit. thanks :)

**Comment by life1347** [comment]
Hi @soamvasani, this PR is ready to merge. Also, I wrote a simple test script under test/mqtrigger to test the functionality of message queue trigger.

File: mqtrigger/messageQueue/nats.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+		// Create a durable subscription to nats, so that triggers could retrieve last unack message.
+		// https://github.com/nats-io/go-nats-streaming#durable-subscriptions
+		ns.DurableName(trigger.Uid),
+
+		// Nats-streaming server is auto-ack mode by default. Since we want nats-streaming server to
+		// resend a message if the trigger does not ack it, we need to enable the manual ack mode, so that
+		// trigger could choose to ack message or simply drop it depend on the response of function pod.
+		ns.SetManualAckMode(),
+	}
+	sub, err := (*nats.nsConn).Subscribe(subj, handler, opts...)
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: Do you need the `(*nats.nsConn)`? Why not just `nats.nsConn`?

File: mqtrigger/messageQueue/nats.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+				log.Warnf("MessageQueue trigger %s deleted failed: %v", trigger.Name, err)
+			} else {
+				log.Infof("MessageQueue trigger %s deleted", trigger.Name)
+			}
+		}
+	}
+}
+
+func isTopicValidForNats(topic string) bool {
+	// nats-streaming does not support wildcard channl.
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: channel*

File: mqtrigger/messageQueue/nats.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+		if resp.StatusCode != 200 {
+			log.Printf("Request returned failure: %v", resp.StatusCode)
+			return
+		}
+		// trigger acks message only if a request done successfully
+		err = msg.Ack()
+		if err != nil {
+			log.Warningf("Failed to ack message: %v", err)
+		}
+		if len(trigger.ResponseTopic) > 0 {
```
Comment (life1347) [inline mqtrigger/messageQueue/nats.go]: I think we should keep the response topic for the following reasons:
1. Nats-Streaming does not provide the request/reply as Nats does. (https://godoc.org/github.com/nats-io/go-nats-streaming#Msg)
2. For the other message queue service, they may not provide the reply topic such as CloudPubSub (https://godoc.org/cloud.google.com/go/pubsub#Message).

File: test/mqtrigger/test.sh [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    exit 1
+fi
+
+cd $FISSIONDIR"/fission/"
+go build
+mv fission $FISSIONDIR"/test/mqtrigger"
+cd $FISSIONDIR"/test/mqtrigger"
+
+./fission env create --name nodejs --image fission/node-env
+./fission fn create --name hello1 --env nodejs --code ~/Desktop/main.js --method GET
```
Comment (life1347) [inline test/mqtrigger/test.sh]: oops

File: mqtrigger/messageQueue/nats.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+const (
+	natsClusterID  = "fissionMQTrigger"
+	natsProtocol   = "nats://"
+	natsClientID   = "fission"
+	natsQueueGroup = "fission-messageQueueNatsTrigger"
+)
+
+type (
+	Nats struct {
+		MessageQueueTriggerManager
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: I don't think you need to wrap the trigger manager into this object. 

You should keep the functionality of managing triggers in the trigger manager, and the functionality of talking to NATS in this file.

File: mqtrigger/messageQueue/nats.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	if !ok {
+		return errors.New("Trigger not exists")
+	}
+	defer nats.delTrigger(triggerUid)
+	nsSub := triggerSub.subscription.(ns.Subscription)
+	err := nsSub.Close()
+	return err
+}
+
+func (nats *Nats) sync() {
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: Is there anything NATS specific here?

File: mqtrigger/messageQueue/nats.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+)
+
+type (
+	Nats struct {
+		MessageQueueTriggerManager
+		nsConn ns.Conn
+	}
+)
+
+func makeNatsTriggerManager(mqTriggerMgr MessageQueueTriggerManager) (MessageQueueTriggerManagerInterface, error) {
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: This should not return a trigger manager interface, but rather a new interface, let's call it `MessageQueue`.

The `MessageQueue` interface would just have simple `subscribe(trigger)` and `unsubscribe(trigger)` methods.  Nats can then satisfy that interface.

File: mqtrigger/messageQueue/nats.go (hunk 7 above)
Comment (life1347) [inline mqtrigger/messageQueue/nats.go]: Originally, I treated Nats as a trigger manager and itself has the functionality of connectint to NATS. But as you said, it's better to keep the functionality in right place instead of putting them together.
After the refactoring, now it's more clear what message queue exactly do and easier to add new message queue in future.
Thanks for the comments, that really helps a lot.

File: mqtrigger/messageQueue/messageQueue.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/fission/fission"
+	controllerClient "github.com/fission/fission/controller/client"
+)
+
+type MessageQueueConfig struct {
```
Comment (soamvasani) [inline mqtrigger/messageQueue/messageQueue.go]: can you keep this in the `type` block below?

File: mqtrigger/messageQueue/messageQueue.go [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+	MQType string
+	Url    string
+}
+
+const (
+	NATS string = "nats-streaming"
+)
+
+type (
+	MessageQueue interface {
```
Comment (soamvasani) [inline mqtrigger/messageQueue/messageQueue.go]: Suggestion -- no need for `triggerSubscription` in this interface. Just return an `interface{}` from `subscribe` and pass the same to `unsubscribe`. The implementations wouldn't use any field other than `triggerSubscription.subscription` anyway.

File: mqtrigger/messageQueue/messageQueue.go [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+		subscribe(trigger fission.MessageQueueTrigger) (*triggerSubscription, error)
+		unsubscribe(triggerSub *triggerSubscription) error
+	}
+
+	MessageQueueTriggerManager struct {
+		sync.RWMutex
```
Comment (soamvasani) [inline mqtrigger/messageQueue/messageQueue.go]: We've so far avoided mutexes for this kind of use case.  Can you change this to use a request and a requestChannel?  See `poolmgr/gpm.go` for an example.

File: mqtrigger/messageQueue/messageQueue.go [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+	var err error
+
+	mqTriggerMgr := MessageQueueTriggerManager{
+		triggerMap: make(map[string]*triggerSubscription),
+		controller: ctrlClient,
+	}
+	switch mqConfig.MQType {
+	case NATS:
+		fallthrough
+	default:
```
Comment (soamvasani) [inline mqtrigger/messageQueue/messageQueue.go]: Why not error on default? Let's keep the choice of message queue explicit.

File: mqtrigger/messageQueue/nats.go [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	triggerSub := triggerSubscription{
+		Metadata: fission.Metadata{
+			Name: trigger.Name,
+			Uid:  trigger.Uid,
+		},
+		funcMeta:     trigger.Function,
+		subscription: sub,
+	}
+	return &triggerSub, nil
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: could just return `sub` here

File: mqtrigger/messageQueue/nats.go [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+func (nats Nats) unsubscribe(triggerSub *triggerSubscription) error {
```
Comment (soamvasani) [inline mqtrigger/messageQueue/nats.go]: suggest changing to `subscription interface{}`

File: publisher/natsPublisher.go [hunk 16]
```diff
@@ -0,0 +1,93 @@
+/*
```
Comment (soamvasani) [inline publisher/natsPublisher.go]: is NatsPublisher used anywhere now? If not, it's still useful, we could put it in a separate pull request.
